cardano-db/src/Cardano/Db/Operations/Delete.hs

```diff
@@ -39,48 +47,69 @@ import Control.Monad.IO.Class (MonadIO, liftIO)
-  delete,
-  deleteWhereCount,
-  selectKeysList,
+  PersistField,
   (!=.),
   (==.),
   (>.),
   (>=.),
  )
+import Database.Persist.Sql (Filter, SqlBackend, delete, deleteWhere, deleteWhereCount, selectKeysList)
 
 deleteBlocksSlotNoNoTrace :: MonadIO m => TxOutTableType -> SlotNo -> ReaderT SqlBackend m Bool
-deleteBlocksSlotNoNoTrace = deleteBlocksSlotNo nullTracer
+deleteBlocksSlotNoNoTrace txOutTableType slotNo = deleteBlocksSlotNo nullTracer txOutTableType slotNo Nothing
 
 -- | Delete a block if it exists. Returns 'True' if it did exist and has been
 -- deleted and 'False' if it did not exist.
-deleteBlocksSlotNo :: MonadIO m => Trace IO Text -> TxOutTableType -> SlotNo -> ReaderT SqlBackend m Bool
-deleteBlocksSlotNo trce txOutTableType (SlotNo slotNo) = do
+deleteBlocksSlotNo ::
+  MonadIO m =>
+  Trace IO Text ->
+  TxOutTableType ->
+  SlotNo ->
+  Maybe Bool ->
+  ReaderT SqlBackend m Bool
+deleteBlocksSlotNo trce txOutTableType (SlotNo slotNo) mIsConsumedTxOut = do
   mBlockId <- queryBlockSlotNo slotNo
   case mBlockId of
-    Nothing -> pure False
-    Just blockId -> do
-      void $ deleteBlocksBlockId trce txOutTableType blockId
+    Nothing -> do
+      liftIO $ logWarning trce $ "deleteBlocksSlotNo: No block contains the the slot: " <> pack (show slotNo)
+      pure False
+    Just (blockId, epochN) -> do
+      void $ deleteBlocksBlockId trce txOutTableType blockId epochN mIsConsumedTxOut
       pure True
 
-deleteBlocksBlockIdNotrace :: MonadIO m => TxOutTableType -> BlockId -> ReaderT SqlBackend m ()
-deleteBlocksBlockIdNotrace txOutTableType = void . deleteBlocksBlockId nullTracer txOutTableType
+deleteBlocksForTests :: MonadIO m => TxOutTableType -> BlockId -> Word64 -> ReaderT SqlBackend m ()
+deleteBlocksForTests txOutTableType blockId epochN = do
+  void $ deleteBlocksBlockId nullTracer txOutTableType blockId epochN Nothing
 
 -- | Delete starting from a 'BlockId'.
-deleteBlocksBlockId :: MonadIO m => Trace IO Text -> TxOutTableType -> BlockId -> ReaderT SqlBackend m (Maybe TxId, Int64)
-deleteBlocksBlockId trce txOutTableType blockId = do
+deleteBlocksBlockId ::
+  MonadIO m =>
+  Trace IO Text ->
+  TxOutTableType ->
+  BlockId ->
+  -- | The 'EpochNo' of the block to delete.
+  Word64 ->
+  -- | Is ConsumeTxout
+  Maybe Bool ->
+  ReaderT SqlBackend m Int64
+deleteBlocksBlockId trce txOutTableType blockId epochN mIsConsumedTxOut = do
   mMinIds <- fmap (textToMinIds txOutTableType =<<) <$> queryReverseIndexBlockId blockId
   (cminIds, completed) <- findMinIdsRec mMinIds mempty
   mTxId <- queryMinRefId TxBlockId blockId
   minIds <- if completed then pure cminIds else completeMinId mTxId cminIds
-  blockCountInt <- deleteTablesAfterBlockId blockId mTxId minIds
-  pure (mTxId, blockCountInt)
+  deleteEpochLogs <- deleteUsingEpochNo epochN
+  (deleteBlockCount, blockDeleteLogs) <- deleteTablesAfterBlockId blockId mTxId minIds
+  whenJust mIsConsumedTxOut $ \_ -> do
+    querySetNullTxOut trce txOutTableType mTxId
```

Comment:
Could we include this in the delete logs summary?
Instead of whenJust here, you could define and use `whenJustDefault`. For inspiration, we have similar utilitis in `Cardano.DbSync.Util.whenDefault,whenMaybe`